### PR TITLE
Condensed linear algebra for device computations

### PR DESCRIPTION
**Pull request by cnpetra**
This PR introduces a condensed KKT reformulation of the XDYcYd linear system, which is then solved using Cholesky factorization and fast sparse matrix-matrix kernels. 

Device computations are supported via the hybrid mode using CUDA (cuSOLVER's low-level Cholesky API and cuSOLVER's factor sparsity-promoting orderings). In CPU mode, MA57 is momentarily used. 

The PR also introduces a CSR storage class and minimal functionality (matrix transpose, matrix-matrix multiplication, and matrix-matrix addition) on host only. Follow-up PRs will "upgrade" this to CSR sparse matrix class and device path computations.

@pelesh @jwang125

### CODE REVIEW

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp [hunk 1]
```diff
@@ -0,0 +1,1245 @@
...
+
+  const index_type* Mirow = M.i_row();
+  const index_type* Mjcol = M.j_col();
+  const double* Mvalues  = M.M();
+
+  //storage the row count
+  index_type w[nrows_];
+  for(int i=0; i<nrows_; ++i) {
+    w[i] = 0;
+  }
```
Comment (ashermancinelli) [inline src/Optimization/hiopKKTLinSysSparseCondensed.cpp:810]:
```suggestion
  //storage the row count
  std::vector<index_type> w(nrows_, 0);
```

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp [hunk 2]
```diff
@@ -0,0 +1,1245 @@
...
+  if(nnz_<=0) {
+    return true;
+  }
+  
+  assert(irowptr_);
+  assert(jcolind_);
+  assert(values_);
+
+  //storage for the row count
+  index_type w[nrows_];
```
Comment (ashermancinelli) [inline src/Optimization/hiopKKTLinSysSparseCondensed.cpp:866]: Same as above for `w`

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp [hunk 3]
```diff
@@ -0,0 +1,1245 @@
...
+
+  irowptr_[0] = 0;
+  for(int i=0; i<nrows_; i++) {
+    irowptr_[i+1] = irowptr_[i] + w[i];
+  }
+  assert(irowptr_[nrows_] == nnz_);
+  return true;
+}
+
+bool hiopMatrixSparseCSRStorage::form_from(const size_type m,
```
Comment (ashermancinelli) [inline src/Optimization/hiopKKTLinSysSparseCondensed.cpp:840]: Seems like this is almost identical to the other `form_from`. Could we just have the `form_from` that takes a matrix just pull the relevant values from the parameter and then call this member function?

**Review by ashermancinelli — APPROVED** [review]
Plenty of nitpicks, but address these as you see fit :)

File: src/LinAlg/hiopLinSolverCholCuSparse.cpp [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+#include "/home/petra1/work/installs/eigen-3.3.9/_install/include/eigen3/Eigen/Core"
+#include "/home/petra1/work/installs/eigen-3.3.9/_install/include/eigen3/Eigen/Sparse"
+
+using Scalar = double;
+//using SparseMatrixCSC = Eigen::SparseMatrix<Scalar, Eigen::StorageOptions::ColMajor>;
+using SparseMatrixCSR = Eigen::SparseMatrix<Scalar, Eigen::StorageOptions::RowMajor>;
+//using Triplet = Eigen::Triplet<Scalar>;
+using Ordering = Eigen::AMDOrdering<SparseMatrixCSR::StorageIndex>;
+using PermutationMatrix = Ordering::PermutationType;
+#endif
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/LinAlg/hiopLinSolverCholCuSparse.cpp:77]: the code is for testing and will be removed once we get to trust cuda sparse orderings more

File: src/LinAlg/hiopLinSolverCholCuSparse.cpp [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+                                            P_h,
+                                            &buf_size);
+    assert(ret == CUSOLVER_STATUS_SUCCESS);
+    
+    // temporary buffer needed for permutation purposes (on host)
+    unsigned char* buf_perm_h = new unsigned char[buf_size];
+    
+    //compute permuted CSR arrays (on host)
+    int* rowptr_perm_h = new int[m+1];
+    int* colind_perm_h = new int[nnz_];
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/LinAlg/hiopLinSolverCholCuSparse.cpp:279]: the code is temporary and will require device allocs soon.

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp [hunk 6]
```diff
@@ -0,0 +1,1245 @@
...
+
+
+  
+  //
+  //update the values in M
+  //
+  double* M_values = M.values();
+  for(int i=0; i<M_nnz; ++i) {
+    M_values[i] = 0.;
+  }
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparseCondensed.cpp:445]: code will be ported to device code soon...

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp (hunk 3 above)
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparseCondensed.cpp:840]: the two methods are a bit different as only one updates the values. The code is temporary and will be refactored when we move to a matrix class.

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp (hunk 1 above)
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparseCondensed.cpp]: fixed

File: src/Optimization/hiopKKTLinSysSparseCondensed.hpp [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+ * (notation) Dd2 = [ (Dd+delta_wd*I)^{-1} + delta_cd*I ]^{-1}
+ *
+ * dd = Jd*dx - delta_cd*dyd - ryd
+ *
+ * From (Dd+delta_wd*I)*dd - dyd = rd_tilde one can write
+ *   ->   (Dd+delta_wd*I)*(Jd*dx - delta_cd*dyd - ryd) - dyd = rd_tilde
+ *   ->   (I+ delta_cd*(Dd+delta_wd*I)) dyd = (Dd+delta_wd*I)*(Jd*dx - ryd) - rd_tilde 
+ * dyd = (I+ delta_cd*(Dd+delta_wd*I))^{-1} [ (Dd+delta_wd*I)*(Jd*dx - ryd) - rd_tilde ]
+ *
+ * (H+Dx+delta_wx*I + Jd^T * Dd2 * Jd) dx = rx_tilde + Jd^T*Dd2*(Dd+delta_wd*I)*ryd +  Jd^T*Dd2*rd_tilde
```
Comment (nychiang) [inline src/Optimization/hiopKKTLinSysSparseCondensed.hpp:191]: I think 
```
Dd2 = [ I + delta_cd*(Dd+delta_wd*I)]^{-1}
and 
(H+Dx+delta_wx*I + Jd^T * Dd2 * (Dd+delta_wd*I) * Jd) dx = rx_tilde + Jd^T*Dd2*(Dd+delta_wd*I)*ryd + Jd^T*Dd2*rd_tilde
```

File: src/Optimization/hiopKKTLinSysSparseCondensed.cpp [hunk 8]
```diff
@@ -0,0 +1,1239 @@
...
+// information, apparatus, product, or process disclosed, or represents that its use would
+// not infringe privately-owned rights.
+// 3. Also, reference herein to any specific commercial products, process, or services by
+// trade name, trademark, manufacturer or otherwise does not necessarily constitute or
+// imply its endorsement, recommendation, or favoring by the United States Government or
+// Lawrence Livermore National Security, LLC. The views and opinions of authors expressed
+// herein do not necessarily state or reflect those of the United States Government or
+// Lawrence Livermore National Security, LLC, and shall not be used for advertising or
+// product endorsement purposes.
+
```
Comment (nychiang) [inline src/Optimization/hiopKKTLinSysSparseCondensed.cpp:47]: add section to record file name and authors

**Review by nychiang — APPROVED** [review]
just one question about the comment about`Dd2`.

File: src/Optimization/hiopKKTLinSysSparseCondensed.hpp (hunk 7 above)
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparseCondensed.hpp:191]: it appears to me that they are equivalent, but your formulation may be better suited for computation (has smaller numbers to invert) and also has a negligible cost overhead...

**Pull request merged**